File: LeetCode75_LevelTwo/House_Robber/House_Robber.cc
// 198. House Robber C++
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    unordered_map<int, int> mp;
    int rob(vector<int>& nums) {
        return robFrom(nums, 0);
    }
    int robFrom(vector<int>& nums, int i) {
        if (i >= nums.size()) return 0;
        if (mp.find(i) != mp.end()) return mp[i];
        int ans = max(robFrom(nums, i + 1), robFrom(nums, i + 2) + nums[i]);
        mp.insert({i, ans});
        return ans;
    }
};

File: LeetCode75_LevelTwo/House_Robber/House_Robber.py
# 198. House Robber Python

class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        self.memo = {}
        return self.robFrom(0, nums)
    def robFrom(self, i, nums):
        if i >= len(nums):
            return 0

        if i in self.memo:
            return self.memo[i]

        output = max(self.robFrom(i + 1, nums), self.robFrom(i + 2, nums) + nums[i])

        self.memo[i] = output
        return output
        